contracts/azorius/LinearERC721Voting.sol

```diff
@@ -0,0 +1,310 @@
+
+        tokenAddresses.push(_tokenAddress);
+        tokenWeights[_tokenAddress] = _weight;
+
+        emit GovernanceTokenAdded(_tokenAddress, _weight);
+    }
+
+    /** @inheritdoc BaseStrategy*/
+    function initializeProposal(bytes memory _data) public override onlyAzorius {
+        uint32 proposalId = abi.decode(_data, (uint32));
+        uint32 _votingEndBlock = uint32(block.number) + votingPeriod;
+
+        proposalVotes[proposalId].votingEndBlock = _votingEndBlock;
+        proposalVotes[proposalId].votingStartBlock = uint32(block.number);
+
+        emit ProposalInitialized(proposalId, _votingEndBlock);
+    }
+
+    /** @inheritdoc BaseStrategy*/
+    function isPassed(uint32 _proposalId) public view override returns (bool) {
+        return (
+            block.number > proposalVotes[_proposalId].votingEndBlock && // voting period has ended
+            quorumThreshold <= proposalVotes[_proposalId].yesVotes + proposalVotes[_proposalId].abstainVotes && // yes + abstain votes meets the quorum
+            meetsBasis(proposalVotes[_proposalId].yesVotes, proposalVotes[_proposalId].noVotes) // yes votes meets the basis
+        );
+    }
+
+    /** @inheritdoc BaseStrategy*/
+    function isProposer(address _address) public view override returns (bool) {
+        uint256 totalWeight = 0;
+        for (uint i = 0; i < tokenAddresses.length;) {
+            address tokenAddress = tokenAddresses[i];
+            totalWeight += IERC721(tokenAddress).balanceOf(_address) * tokenWeights[tokenAddress];
+            unchecked { ++i; }
+        }
+        return totalWeight >= proposerThreshold;
+    }
+
+    /** @inheritdoc BaseStrategy*/
+    function votingEndBlock(uint32 _proposalId) public view override returns (uint32) {
+      return proposalVotes[_proposalId].votingEndBlock;
+    }
+
+    // verifies the voter holds the NFTs and returns the total weight associated with their tokens
+    // the frontend will need to determine whether an address can vote on a proposal, as it is possible
+    // to vote twice if you get more weight later on
+    function _getTotalWeight(
+        uint256 _proposalId,
+        address[] memory _tokenAddresses,
+        uint256[] memory _tokenIds,
+        address _voter
+    ) internal returns (uint256) {
+
+        uint256 weight = 0;
+
+        for (uint256 i = 0; i < _tokenAddresses.length; i++) {
+
+            address tokenAddress = _tokenAddresses[i];
+            uint256 tokenId = _tokenIds[i];
+
+            // ensure the token hasn't voted already, and the voter actually holds the token
+            if (
+                proposalVotes[_proposalId].hasVoted[tokenAddress][tokenId] == true || 
+                _voter != IERC721(tokenAddress).ownerOf(tokenId)
+            ) {
+                continue;
+            }
+            
+            weight += tokenWeights[tokenAddress];
+            proposalVotes[_proposalId].hasVoted[tokenAddress][tokenId] = true;
+        }
```

Comment:
Thoughts on moving this loop into the _vote function? It does seem confusing to me that a function called `_getTotalWeight` isn't a view function, and is changing state